### PR TITLE
modify interface for Consensus module && implement the base functions for consensus module && modify Block and BlockHeader to meet the need of consensus module

### PR DESCRIPTION
**Pull request by cyjseagull**
1. modify interface for Consensus module && implement the base functions for consensus module
2. modify BlockManager interface to add getLatestBlockHeader interface
3. modify Block and BlockHeader to meet the need of consensus module

### CODE REVIEW

File: libconsensus/Common.h [hunk 1]
```diff
@@ -21,12 +21,21 @@
...
 #pragma once
+#include <libdevcore/FixedHash.h>
 namespace dev
 {
 namespace consensus
 {
-enum ConsensusStatus
+enum NodeAccountType
 {
+    NonMinerAccount = 0,
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Common.h]: already modified, please refer to : https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-9cb3c483050ad9fd820030d43705bde7R36

File: libconsensus/Consensus.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            LOG(INFO) << "Skipping txpool sync for a sealed block";
+            return;
+        }
+        /// sealed transaction over maxBlockTransactions
+        if (m_sealing.getTransactionSize() >= maxBlockTransactions)
+        {
+            LOG(INFO) << "Skipping txq sync for full block";
+            return;
+        }
+        m_sealing.appendTransactions(m_txPool->topTransactions(maxBlockTransactions));
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-db2750f4390ce4f60b54c9d735c632d7R97 and https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-2ea0349e28a58ead9f13cc24ba1288deR244

File: libconsensus/Consensus.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+void Consensus::doWork(bool wait)
+{
+    bool sealed = false;
+    bool condition = true;
+    DEV_READ_GUARDED(x_sealing)
+    sealed = m_sealing.isSealed();
+    /// TODO: m_remoteWorking
+    bool t = false;
+    if (!sealed && !m_blockSync->isSyncing() && m_syncTxPool.compare_exchange_strong(t, false))
+        syncTxPool(m_maxBlockTransactions);
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: already removed

[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: `loadTransactions` has been added, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-db2750f4390ce4f60b54c9d735c632d7R80

File: libconsensus/Consensus.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        return;
+    }
+    doWork(false);
+    startWorking();
+}
+
+/// doWork
+void Consensus::doWork(bool wait)
+{
+    bool sealed = false;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: checking condition has been modified, please refer to： https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-db2750f4390ce4f60b54c9d735c632d7R51

File: libconsensus/Consensus.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    bool condition = true;
+    DEV_READ_GUARDED(x_sealing)
+    sealed = m_sealing.isSealed();
+    /// TODO: m_remoteWorking
+    bool t = false;
+    if (!sealed && !m_blockSync->isSyncing() && m_syncTxPool.compare_exchange_strong(t, false))
+        syncTxPool(m_maxBlockTransactions);
+    /// sealing block
+    rejigSealing();
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: already moved.

File: libconsensus/Consensus.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+{
+/// start the consensus module
+void Consensus::start()
+{
+    if (m_startSeal)
+    {
+        LOG(WARNING) << "Consensus module has already been started, return directly";
+        return;
+    }
+    doWork(false);
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: removed

File: libconsensus/Consensus.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    startWorking();
+}
+
+/// doWork
+void Consensus::doWork(bool wait)
+{
+    bool sealed = false;
+    bool condition = true;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: removed

File: libconsensus/Consensus.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+using namespace dev::sync;
+using namespace dev::blockmanager;
+namespace dev
+{
+namespace consensus
+{
+/// start the consensus module
+void Consensus::start()
+{
+    if (m_startSeal)
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: modified to m_startConsensus, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-db2750f4390ce4f60b54c9d735c632d7R157

File: libconsensus/Consensus.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    /// set transaction root, receipt root and state root
+    m_sealingHeader.setRoots(
+        m_sealing.getTransactionRoot(), hash256(BytesMap()), hash256(BytesMap()));
+    m_sealingHeader.setExtraData(m_extraData);
+}
+
+/// sealing function
+bool Consensus::generateSeal(bytes& _block_data)
+{
+    if (m_startSeal && getNodeAccountType() == NodeAccountType::MinerAccount && !isBlockSyncing() &&
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: modified, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/75/files#diff-db2750f4390ce4f60b54c9d735c632d7R51

File: libconsensus/Consensus.cpp [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        if (tx_num < max_block_can_seal && !m_blockSync->isSyncing() &&
+            m_syncTxPool.compare_exchange_strong(t, false))
+            loadTransactions(tx_num, max_block_can_seal);
+        if (shouldWaitForNextInterval())
+        {
+            LOG(DEBUG) << "Wait for next interval, tx:" << m_sealing.getTransactionSize();
+            return false;
+        }
+        submitToEncode();
+        m_sealingHeader = m_sealing.blockHeader();
```
Comment (wheatli) [inline libconsensus/Consensus.cpp]: not essential

File: libconsensus/Consensus.cpp [hunk 11]
```diff
@@ -40,106 +40,100 @@ void Consensus::start()
...
     resetCurrentTime();
-    m_sealingHeader.setSealerList(minerList());
-    m_sealingHeader.setSealer(nodeIdx());
-    m_sealingHeader.setLogBloom(LogBloom());
-    m_sealingHeader.setGasUsed(0);
+    m_sealing.sealing_block.header().setSealerList(minerList());
+    m_sealing.sealing_block.header().setSealer(nodeIdx());
+    m_sealing.sealing_block.header().setLogBloom(LogBloom());
+    m_sealing.sealing_block.header().setGasUsed(0);
+    m_sealing.sealing_block.header().setExtraData(m_extraData);
```
Comment (wheatli) [inline libconsensus/Consensus.cpp]: TODO:  We need a ExxcutiveContext here to get the result of txs executed, such as state_root, receipt_root and so on

File: libconsensus/Consensus.h [hunk 12]
```diff
@@ -36,6 +36,11 @@ namespace dev
 class ConsensusStatus;
 namespace consensus
 {
+struct Sealing
+{
+    Block sealing_block;
```
Comment (wheatli) [inline libconsensus/Consensus.h]: sealing_block -> block

File: libtxpool/TxPool.cpp [hunk 13]
```diff
@@ -241,23 +243,28 @@ bool TxPool::drop(h256 const& _txHash)
...
-            if (!_avoid.count(it->sha3()))
-                ret.push_back(*it);
+            ret.push_back(*it);
+            m_transactionSet.insert(it->sha3());
         }
     }
     return ret;
 }
 
+/// obtain transactions with filter logic
```
Comment (wheatli) [inline libtxpool/TxPool.cpp]: not essential

File: libtxpool/TxPool.h [hunk 14]
```diff
@@ -158,6 +157,8 @@ class TxPool : public TxPoolInterface, public std::enable_shared_from_this<TxPoo
     h256Hash m_known;
     /// hash of dropped transactions
     h256Hash m_dropped;
+    /// hash set for filter fetched transactions
+    h256Hash m_transactionSet;
```
Comment (wheatli) [inline libtxpool/TxPool.h]: not essential

File: libconsensus/Consensus.h [hunk 15]
```diff
@@ -159,6 +166,8 @@ class Consensus : public Worker, virtual ConsensusInterface
 
     NodeAccountType m_accountType;
     u256 m_idx = 0;
+    /// hash set for filter fetched transactions
+    h256Hash m_transactionSet;
```
Comment (wheatli) [inline libconsensus/Consensus.h]: place into struct Sealing

File: libconsensus/Consensus.cpp (hunk 11 above)
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: This work is done by: setSealingRoot

File: libconsensus/Consensus.cpp [hunk 16]
```diff
@@ -40,106 +40,100 @@ void Consensus::start()
...
-        bool ret = generateBlock(block_data, tx_num, max_block_can_seal);
-        if (ret == true)
+        DEV_WRITE_GUARDED(x_sealing)
         {
-            handleBlock(m_sealingHeader, ref(block_data));
+            /// get current transaction num
+            uint64_t tx_num = m_sealing.sealing_block.getTransactionSize();
+            /// obtain the transaction num should be packed
+            uint64_t max_blockCanSeal = calculateMaxPackTxNum();
+            if (max_blockCanSeal < tx_num)
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Consensus.cpp]: done

**Pull request merged**